cachi2/core/package_managers/yarn/main.py

```diff
@@ -0,0 +1,84 @@
+import logging
+
+from cachi2.core.models.input import Request
+from cachi2.core.models.output import Component, EnvironmentVariable, RequestOutput
+from cachi2.core.package_managers.yarn.project import Project, YarnRc
+from cachi2.core.package_managers.yarn.resolver import (
+    create_component_from_package,
+    resolve_dependency_info,
+)
+from cachi2.core.package_managers.yarn.utils import run_yarn_cmd
+from cachi2.core.rooted_path import RootedPath
+
+log = logging.getLogger(__name__)
+
+
+def fetch_yarn_source(request: Request) -> RequestOutput:
+    """Process all the yarn source directories in a request."""
+    components = []
+
+    for package in request.yarn_packages:
+        path = request.source_dir.join_within_root(package.path)
+        project = Project.from_source_dir(path)
+        components.extend(_resolve_yarn_project(project, request.output_dir))
+
+    return RequestOutput.from_obj_list(
+        components, _generate_environment_variables(), project_files=[]
+    )
+
+
+def _resolve_yarn_project(project: Project, output_dir: RootedPath) -> list[Component]:
+    """Process a request for a single yarn source directory.
+
+    :param project: the directory to be processed.
+    :param output_dir: the directory where the prefetched dependencies will be placed.
+    """
+    _configure_yarn_version(project)
+
+    try:
+        _prepare_for_prefetch(project.yarn_rc, output_dir)
+        packages = resolve_dependency_info(project.source_dir)
+        _fetch_dependencies(project)
+    except Exception:
+        _restore_files(project)
+        raise
+
+    _restore_files(project)
+
+    return [create_component_from_package(package, project.yarn_rc) for package in packages]
+
+
+def _configure_yarn_version(project: Project) -> None:
+    """Resolve the yarn version and set it in the package.json file if needed.
+
+    :raises PackageRejected: in case the yarn version can't be determined, or if
+    there is a mismatch between the version in package.json and yarnrc.yml.
+    """
+    pass
+
+
+def _prepare_for_prefetch(yarnrc: YarnRc, output_dir: RootedPath) -> None:
+    """Change the configuration in yarnrc to perform the prefetch.
+
+    :param output_dir: the directory where the prefetched dependencies will be placed.
+    """
+    pass
+
+
+def _fetch_dependencies(project: Project) -> None:
+    """Fetch dependencies using 'yarn install'.
+
+    :raises SubprocessCallError: if the 'yarn install' command fails.
+    """
+    run_yarn_cmd(["yarn", "--help"], project.source_dir)
+    pass
+
+
+def _restore_files(project: Project) -> None:
+    """Restore the files that were changed during the prefetch."""
+    pass
+
+
+def _generate_environment_variables() -> list[EnvironmentVariable]:
+    """Generate environment variables that will be used for building the repository that is being processed."""
+    return NotImplemented
```

Comment:
Would we want to set YARN_GLOBAL_MIRROR even if all the projects are zero-installs and $output_dir/deps/yarnberry is empty? (If that's how we handle zero-installs)? I guess it doesn't really hurt anything

Reply:
Do you mean YARN_ENABLE_MIRROR? It probably doesn't hurt (we will probably mount the ouput_dir in RHTAP regardless), but I think it is also easy enough to pass the project variable here and set it accordingly.

Reply:
Both YARN_GLOBAL_MIRROR and YARN_ENABLE_MIRROR

We will mount the output dir at /cachi2, /cachi2/deps/yarnberry/ will be empty or nonexistent. We may or may not set `YARN_GLOBAL_MIRROR=/cachi2/deps/yarnberry` and `YARN_ENABLE_MIRROR=true`

Depending on what the user does in the Dockerfile, they might end up writing to / reading from the global mirror. Yarn should do the right thing when reading from an empty/non-existent global mirror. If Yarn writes something to the global mirror, that should be fine too - by that point, RHTAP will hopefully have copied the things it needs for the source container.

Reply:
But anyway, from a high-level code design perspective, the only thing that would need to change is passing the Project :+1:

Reply:
(And making sure that `Project.is_zero_installs` does the right thing even if you've already called `yarn install`)